fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java

```diff
@@ -317,6 +329,21 @@ public <R> CompletableFuture<R> apply(Function<FDBRecordContext, CompletableFutu
             });
         }
 
+        @Override
+        @SuppressWarnings("PMD.CloseResource") // closed in a future
+        public <R> CompletableFuture<R> applyInRecoveryPath(Function<FDBRecordContext, CompletableFuture<R>> function) {
+            // Create a new, dedicated context. Apply, flush, and close it.
+            FDBRecordContextConfig contextConfig = contextConfigBuilder.build();
+            final FDBRecordContext recoveryContext = database.openContext(contextConfig);
+            return function.apply(recoveryContext)
```

Comment:
if `function` throws an exception, you won't ever close the `recoveryContext`.
The easiest way I've found to do this is something like:
```
boolean successful = false;
try {
    future = function.apply(recoveryContext).whenComplete( /* ... as you have it */);
    successful = true;
} finally {
    if (!successful) {
    recoveryContext.close();
}
```

Reply:
✅